### ISSUE TITLE
Termination -- java.lang.IllegalArgumentException

### ISSUE THREAD

**Issue by BLepers** [epfl-lara/leon#213]
Tested with 9566dcff6d0ee365565d9da635d5c463ce943935

With [Order.scala](https://github.com/epfl-lara/leon/files/267687/Order.txt) (the file is not minimal, but hopefully not too hard to debug :) ) The problem is raised by the termination verification of testEvolutionSimple.

```
../leon/leon --termination Order.scala 
java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.TimSort.mergeHi(TimSort.java:899)
	at java.util.TimSort.mergeAt(TimSort.java:516)
	at java.util.TimSort.mergeForceCollapse(TimSort.java:457)
	at java.util.TimSort.sort(TimSort.java:254)
	at java.util.Arrays.sort(Arrays.java:1438)
	at scala.collection.SeqLike$class.sorted(SeqLike.scala:648)
	at scala.collection.AbstractSeq.sorted(Seq.scala:41)
	at scala.collection.SeqLike$class.sortWith(SeqLike.scala:601)
	at scala.collection.AbstractSeq.sortWith(Seq.scala:41)
	at leon.termination.Strengthener$class.strengthenPostconditions(Strengthener.scala:25)
	at leon.termination.ComplexTerminationChecker$$anon$3.strengthenPostconditions(ComplexTerminationChecker.scala:34)
	at leon.termination.RelationProcessor.run(RelationProcessor.scala:21)
	at leon.termination.ProcessingPipeline$$anon$1.next(ProcessingPipeline.scala:189)
	at leon.termination.ProcessingPipeline$$anon$1.next(ProcessingPipeline.scala:182)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:462)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at leon.termination.ProcessingPipeline.verifyTermination(ProcessingPipeline.scala:216)
	at leon.termination.ProcessingPipeline$$anonfun$5$$anonfun$apply$4.apply(ProcessingPipeline.scala:144)
	at leon.termination.ProcessingPipeline$$anonfun$5$$anonfun$apply$4.apply(ProcessingPipeline.scala:143)
	at scala.Option.getOrElse(Option.scala:121)
	at leon.termination.ProcessingPipeline$$anonfun$5.apply(ProcessingPipeline.scala:142)
	at leon.termination.ProcessingPipeline$$anonfun$5.apply(ProcessingPipeline.scala:133)
	at scala.Option.getOrElse(Option.scala:121)
	at leon.termination.ProcessingPipeline.terminates(ProcessingPipeline.scala:133)
	at leon.termination.TerminationPhase$$anonfun$4.apply(TerminationPhase.scala:29)
	at leon.termination.TerminationPhase$$anonfun$4.apply(TerminationPhase.scala:28)
	at scala.collection.immutable.List.map(List.scala:277)
	at leon.termination.TerminationPhase$.apply(TerminationPhase.scala:28)
	at leon.termination.TerminationPhase$.apply(TerminationPhase.scala:8)
	at leon.SimpleLeonPhase$class.run(LeonPhase.scala:14)
	at leon.termination.TerminationPhase$.run(TerminationPhase.scala:8)
	at leon.Pipeline$$anon$1.run(Pipeline.scala:10)
	at leon.Pipeline$$anon$1.run(Pipeline.scala:12)
	at leon.Main$.execute(Main.scala:267)
	at leon.Main$.main(Main.scala:251)
	at leon.Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at scala.reflect.internal.util.ScalaClassLoader$$anonfun$run$1.apply(ScalaClassLoader.scala:70)
	at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:101)
	at scala.reflect.internal.util.ScalaClassLoader$class.run(ScalaClassLoader.scala:70)
	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:101)
	at scala.tools.nsc.CommonRunner$class.run(ObjectRunner.scala:22)
	at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:39)
	at scala.tools.nsc.CommonRunner$class.runAndCatch(ObjectRunner.scala:29)
	at scala.tools.nsc.ObjectRunner$.runAndCatch(ObjectRunner.scala:39)
	at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:65)
	at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:87)
	at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:98)
	at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:103)
	at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```

**Comment by samarion** [comment]
Thanks!
Should be fixed in a679d54b7cf4634c9c0874f00fac82d3ead08b8b.

I can't reproduce it on my machine (probably hardware dependent ordering after toSeq) so I would appreciate a final check. Please reopen if it still fails.

**Comment by BLepers** [comment]
Thanks, it seems to work now :)

**Comment by celia01** [comment]
> Thanks!
> Should be fixed in [a679d54](https://github.com/epfl-lara/leon/commit/a679d54b7cf4634c9c0874f00fac82d3ead08b8b).
> 
> I can't reproduce it on my machine (probably hardware dependent ordering after toSeq) so I would appreciate a final check. Please reopen if it still fails.

Excuse me, Did you know how to solve this problem with using sortWith rather than sorted？

I solved this TimSort problem by replacing sortWith with sortBy，too.
But I often use sortWith in my code。
However，the return type of comparator in sortWith is Bool, so I want to know how to use sortWith to solve the problem?
